### PR TITLE
Cortx 29556 iam users rest api

### PR DESCRIPTION
**Pull request by amit-si-seagate**
# Problem Statement
- I AM user API. (Create, Get, delete, add/remove user keys, modify user keys and add/remove caps). [Cortx 29556] 

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: libs/csm/rest/csm_rest_iamuser.py [hunk 1]
```diff
@@ -587,7 +587,47 @@ def get_iam_user_rgw(self, uid, header):
                                           headers=header)
         self.log.info("Get IAM user request successfully sent...")
         return response
+    
+    @RestTestLib.authenticate_and_login
+    def get_iam_user_payload_rgw(self, payload: dict, auth_header=None):
```
Comment (kachhwahadivya) [inline libs/csm/rest/csm_rest_iamuser.py:592]: These functions can be clubbed to existing functions with auth_header flag where ever it is not there.

File: tests/csm/rest/test_iam_users.py [hunk 2]
```diff
@@ -386,7 +409,1460 @@ def test_35607(self):
         self.log.info("TODO Verify Response : %s", resp)
         self.log.info("[END]Creating IAM user with basic parameters")
         self.log.info("##### Test completed -  %s #####", test_case_name)
+    
+    ###
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_iam_users.py:413]: This can be removed. Also Add @pytest.mark.lc

File: tests/csm/rest/test_iam_users.py [hunk 3]
```diff
@@ -386,7 +409,1460 @@ def test_35607(self):
...
+        self.log.info("Perform API to Create IAM User with already existing user Access Keys.")
+        user_id3, display_name3, email3 = self.get_IAM_user_payload("email")
+        payload = {"uid": user_id3, "display_name": display_name3, "email": email3,
+                   "access_key": resp["keys"][0]["access_key"], "secret_key": resp["keys"][0]["secret_key"]}
+
+        self.log.info("payload:  %s", payload)
+        self.log.info("Verify keys_exist error)")
+        resp_new = self.csm_obj.create_iam_user_rgw(payload)
+        assert resp_new.status_code == HTTPStatus.CONFLICT, "Check failed for duplicate user creation"
+        self.log.info("[END] Testing with duplicate uid and email")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_iam_users.py:458]: Suggestion added completed statement as it is easier to debug
        self.log.info("##### Test ended -  %s #####", test_case_name)

File: tests/csm/rest/test_iam_users.py [hunk 4]
```diff
@@ -386,7 +409,1460 @@ def test_35607(self):
...
+        payload = {"user_caps":"usage=read,write;user=write"}
+        response = self.csm_obj.add_user_caps_rgw(user_id, payload)
+        assert response.status_code == HTTPStatus.OK, "Status code check failed for add capability."
+        
+        self.log.info("Perform PUT request to remove invalid capability")
+        payload = {"user_caps":"random=*;buckets=*"}
+        response = self.csm_obj.remove_user_caps_rgw(user_id, payload)
+        assert response.status_code == HTTPStatus.BAD_REQUEST, "Status code check failed for add capability."
+        
+        self.log.info("Perform PUT request to remove invalid  caps value")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_iam_users.py:1853]: Is this correct log for below statement?

File: tests/csm/rest/test_iam_users.py [hunk 5]
```diff
@@ -386,7 +409,1460 @@ def test_35607(self):
...
+        payload = {"user_caps":"users=random;buckets=*"}
+        response = self.csm_obj.remove_user_caps_rgw(user_id, payload)
+        assert response.status_code == HTTPStatus.OK, "Status code check failed for add capability."
+
+        self.log.info("Perform PUT request to remove invalid  format caps")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_iam_users.py:1858]: Is below capability invalid?

File: libs/csm/rest/csm_rest_iamuser.py (hunk 1 above)
Comment (seagate-sarang-sawant) [reply libs/csm/rest/csm_rest_iamuser.py:592]: @amit-si-seagate  I agree with @kachhwahadivya  we needs to have only 1 interfacting API function to be consumed rom out of CSM components. The new added methods if they don't exists can become private/protected methods (prefix  __ or  _) .

Comment (seagate-sarang-sawant) [reply libs/csm/rest/csm_rest_iamuser.py:592]: This comment applies to all newly added methods.

**Review by seagate-sarang-sawant — CHANGES_REQUESTED** [review]

File: config/csm/csm_config.yaml [hunk 6]
```diff
@@ -26,6 +26,8 @@ Restcall:
   s3_bucket_endpoint : "/api/v2/s3/bucket"
   IAM_users_endpoint : "/api/v2/iam_users"
   s3_iam_user_endpoint : "/api/v2/s3/iam/users"
+  s3_iam_user_keys_endpoint : "/api/v2/s3/iam/keys"
+  s3_iam_caps_endpoint:  "/api/v2/s3/iam/caps"
```
[the remark this answers is not in the document]
Comment (kachhwahadivya) [reply config/csm/csm_config.yaml:30]: @seagate-sarang-sawant : We had 2 different instances as they might change version later. So we want to keep both of them separate until LR and LC are merged to same branch.

File: config/csm/csm_config.yaml [hunk 7]
```diff
@@ -26,6 +26,8 @@ Restcall:
   s3_bucket_endpoint : "/api/v2/s3/bucket"
   IAM_users_endpoint : "/api/v2/iam_users"
   s3_iam_user_endpoint : "/api/v2/s3/iam/users"
+  s3_iam_keys_endpoint : "/api/v2/s3/iam/keys"
+  s3_iam_caps_endpoint : "/api/v2/s3/iam/caps"
```
Comment (kachhwahadivya) [inline config/csm/csm_config.yaml:30]: This endpoints are not applicable for LR, So please remove them from here and use it from Restcall_LC

File: libs/csm/rest/csm_rest_iamuser.py [hunk 8]
```diff
@@ -556,7 +553,9 @@ def remove_user_caps_rgw(self, uid, payload: dict):
         self.log.info("Removing user capabilities from user request successfully sent...")
         return response
 
-    def delete_iam_user_rgw(self, uid, header, purge_data=False):
+    @RestTestLib.authenticate_and_login
+    def delete_iam_user_rgw(self, uid, header, purge_data=False, payload: dict = {},
```
Comment (kachhwahadivya) [inline libs/csm/rest/csm_rest_iamuser.py:557]: Do we need payload here?

File: libs/csm/rest/csm_rest_iamuser.py (hunk 1 above)
Comment (amit-si-seagate) [reply libs/csm/rest/csm_rest_iamuser.py]: Resolved @kachhwahadivya

File: tests/csm/rest/test_iam_users.py (hunk 2 above)
Comment (amit-si-seagate) [reply tests/csm/rest/test_iam_users.py]: Resolved cc @kachhwahadivya

File: tests/csm/rest/test_iam_users.py (hunk 3 above)
Comment (amit-si-seagate) [reply tests/csm/rest/test_iam_users.py:450]: Resolved @kachhwahadivya

File: tests/csm/rest/test_iam_users.py (hunk 4 above)
Comment (amit-si-seagate) [reply tests/csm/rest/test_iam_users.py]: Resolved.

File: tests/csm/rest/test_iam_users.py (hunk 5 above)
Comment (amit-si-seagate) [reply tests/csm/rest/test_iam_users.py]: Resolved.

File: config/csm/csm_config.yaml (hunk 7 above)
Comment (amit-si-seagate) [reply config/csm/csm_config.yaml]: Resolved.

File: libs/csm/rest/csm_rest_iamuser.py (hunk 8 above)
Comment (amit-si-seagate) [reply libs/csm/rest/csm_rest_iamuser.py]: Resolved

**Review by kachhwahadivya — APPROVED** [review]

**Pull request closed without merging**